lib/backend/kubernetes/kubernetes.go

```diff
@@ -0,0 +1,263 @@
+}
+
+// Backend uses Kubernetes Secrets to store identities.
+type Backend struct {
+	// kubernetes client
+	k8sClientSet *kubernetes.Clientset
+	namespace    string
+	secretName   string
+	replicaName  string
+
+	// Mutex is used to limit the number of concurrent operations per agent to 1 so we do not need
+	// to handle retries locally.
+	// The same happens with SQlite backend.
+	mu *sync.Mutex
+}
+
+// New returns a new instance of Kubernetes Secret identity backend storage.
+func New() (*Backend, error) {
+
+	restClient, _, err := kubeutils.GetKubeClient("")
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &Backend{
+		k8sClientSet: restClient,
+		namespace:    os.Getenv(namespaceEnv),
+		replicaName:  os.Getenv(teleportReplicaNameEnv),
+		secretName: fmt.Sprintf(
+			"%s-%s",
+			os.Getenv(teleportReplicaNameEnv),
+			secretIdentifierName,
+		),
+		mu: &sync.Mutex{},
+	}, nil
+}
+
+// Exists checks if the secret already exists in Kubernetes.
+// It's used to determine if the agent never created a secret and might upgrade from
+// local SQLite database. In that case, the agent reads local database and
+// creates a copy of the keys in Kube Secret.
+func (b *Backend) Exists(ctx context.Context) bool {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	_, err := b.getSecret(ctx)
+	return err == nil
+}
+
+// Get reads the secret and extracts the key from it.
+// If the secret does not exist or the key is not found it returns trace.Notfound,
+// otherwise returns the underlying error.
+func (b *Backend) Get(ctx context.Context, key []byte) (*backend.Item, error) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	return b.readSecretData(ctx, key)
+}
+
+// Create creates item
+func (b *Backend) Create(ctx context.Context, i backend.Item) (*backend.Lease, error) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	return b.updateSecretContent(ctx, i)
+}
+
+// Put puts value into backend (creates if it does not exist, updates it otherwise)
+func (b *Backend) Put(ctx context.Context, i backend.Item) (*backend.Lease, error) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	return b.updateSecretContent(ctx, i)
+}
+
+// PutItems receives multiple items and upserts them into the Kubernetes Secret.
+// This function is only used when the Agent's Secret does not exist, but local SQLite database
+// has identity credentials.
+// TODO(tigrato): remove this once the compatibility layer between local storage and
+// Kube secret storage is no longer required!
+func (b *Backend) PutItems(ctx context.Context, items ...backend.Item) error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	_, err := b.updateSecretContent(ctx, items...)
+	return err
+}
+
+// getSecret reads the secret from K8S API.
+func (b *Backend) getSecret(ctx context.Context) (*corev1.Secret, error) {
+	secret, err := b.k8sClientSet.
+		CoreV1().
+		Secrets(b.namespace).
+		Get(ctx, b.secretName, metav1.GetOptions{})
+
+	if err == nil {
+		return secret, nil
+	}
+	kubeErr := &kubeerrors.StatusError{}
```

Comment:
nit: doesn't this also work? I feel like it's more idiomatic
```suggestion
	var kubeErr kubeerrors.StatusError
```

Reply:
error interface is implemented for `*kubeerrors.StatusError`, and we need to call `errors.As` with a pointer to an object that implements the error interface

Reply:
Got it